enterprise/server/remote_execution/snaploader/snaploader_test.go

```diff
@@ -81,67 +230,68 @@ func TestPackAndUnpackChunkedFiles(t *testing.T) {
+	cowA, err := copy_on_write.ConvertFileToCOW(ctx, env, originalImagePath, chunkSize, workDirA)
 	require.NoError(t, err)
-	// Overwrite a random range to simulate the disk being written to. This
-	// should create some dirty chunks.
 	writeRandomRange(t, cowA)
-	// Now store a snapshot for VM A, including the COW we created.
-	taskA := &repb.ExecutionTask{}
-	keyA, err := snaploader.NewKey(taskA, "config-hash-a", "")
+	task := &repb.ExecutionTask{}
+	key, err := snaploader.NewKey(task, "config-hash", "")
 	require.NoError(t, err)
 	optsA := makeFakeSnapshot(t, workDirA)
 	optsA.ChunkedFiles = map[string]*copy_on_write.COWStore{
 		"scratchfs": cowA,
 	}
-	err = loader.CacheSnapshot(ctx, keyA, optsA)
-	require.NoError(t, err)
-	// Note: we'd normally close cowA here, but we keep it open so that
-	// mustUnpack() can verify the original contents.
-
-	// Read the bytes from cowA now to avoid relying on cowA being immutable
-	// (though it should be).
-	scratchfsBytesA := mustReadStore(t, cowA)
-
-	// Now unpack the snapshot for use by VM B, then make a modification.
-	workDirB := testfs.MakeDirAll(t, workDir, "VM-B")
-	unpackedB := mustUnpack(t, ctx, loader, keyA, workDirB, optsA)
-	cowB := unpackedB.ChunkedFiles["scratchfs"]
-	writeRandomRange(t, cowB)
-
-	// Unpack the snapshot again for use by VM C. It should see the original
-	// contents, not the modification made by VM B.
-	workDirC := testfs.MakeDirAll(t, workDir, "VM-C")
-	unpackedC := mustUnpack(t, ctx, loader, keyA, workDirC, optsA)
-	// mustUnpack already verifies the contents against cowA, but this will give
-	// us false confidence if cowA was somehow mutated. So check again against
-	// the originally snapshotted bytes, rather than the original COW instance.
-	r, err := interfaces.StoreReader(unpackedC.ChunkedFiles["scratchfs"])
-	require.NoError(t, err)
-	scratchfsBytesC, err := io.ReadAll(r)
-	require.NoError(t, err)
-	if !bytes.Equal(scratchfsBytesA, scratchfsBytesC) {
-		require.FailNow(t, "scratchfs bytes for VM C should match original contents from snapshot A")
+	err = loader.CacheSnapshot(ctx, key, optsA)
+	require.NoError(t, err)
+
+	// Delete some artifacts from the remote cache
+	snapMetadata, err := loader.GetSnapshot(ctx, key)
+	require.NoError(t, err)
+	for i, f := range snapMetadata.GetFiles() {
+		if i == 0 {
+			rn := digest.NewResourceName(f.GetDigest(), "", rspb.CacheType_CAS, repb.DigestFunction_BLAKE3).ToProto()
+			err = env.GetCache().Delete(ctx, rn)
+			require.NoError(t, err)
+			break
+		}
+	}
+	for _, f := range snapMetadata.GetChunkedFiles() {
+		for i, c := range f.GetChunks() {
+			if i == 0 {
+				rn := digest.NewResourceName(c.GetDigest(), "", rspb.CacheType_CAS, repb.DigestFunction_BLAKE3).ToProto()
+				err = env.GetCache().Delete(ctx, rn)
+				require.NoError(t, err)
+				break
+			}
+		}
 	}
+
+	// Even though the assets still exist in the local cache, the remote cache
```

Comment:
Is this the desired behavior? My understanding coming out of our meeting was that we wanted one source of truth for what snapshot VMs should be using

Reply:
I think it's fine to keep it like this for now and then see how it performs in dev.

I'm betting that we're going to need some sort of configurable "max age" value after which snapshots have to be re-fetched from the remote cache, like 3 hours or something. I think this is mainly going to be needed for users that don't keep their build graphs super lean, which is probably pretty common. Our build graph for example is relatively heavyweight since we only have an 89% AC hit rate, so I suspect we'll have ~gigabytes of cache churn on every commit pushed to master.

Maybe we should start storing the snapshot creation time somewhere in the snapshot metadata, so that we can see how much of a correlation there is between snapshot age and workflow execution time.

Reply:
Storing snapshot age is a good idea. I can look into that when we start to do more performance improvements